addons/binding/org.openhab.binding.bigassfan/src/main/java/org/openhab/binding/bigassfan/handler/BigAssFanHandler.java

```diff
@@ -0,0 +1,1030 @@
+                }
+                return null;
+            }
+            return readMessage();
+        }
+
+        private String readMessage() {
+            logger.debug("Waiting for message from fan at {}", ipAddress);
+            String message = conn.read();
+            logger.debug("FanListener received message of length {}: {}", message.length(), message);
+            return message;
+        }
+
+        private void processMessage(String message) {
+            if (StringUtils.isEmpty(message)) {
+                return;
+            }
+
+            // Match on (msg)
+            logger.info("FanListener processing received message {}", message);
+            Pattern pattern = Pattern.compile("[(](.*)");
+            Matcher matcher = pattern.matcher(message);
+            if (!matcher.find()) {
+                logger.debug("Unable to process message, doesn't match expected format: {}", message);
+                return;
+            }
+
+            message = matcher.group(1);
+            String[] messageParts = message.split(";");
+
+            // Check to make sure it is my MAC address or my label
+            if (!isMe(messageParts[0])) {
+                logger.trace("Message not for me: {} - {}", messageParts[0], macAddress);
+                return;
+            }
+
+            logger.debug("Message is for me!!!  Process the message: {}", message);
+            String messageUpperCase = message.toUpperCase();
+            if (messageUpperCase.contains(";FAN;PWR;")) {
+                updateFanPower(messageParts);
+
+            } else if (messageUpperCase.contains(";FAN;SPD;ACTUAL;")) {
+                updateFanSpeed(messageParts);
+
+            } else if (messageUpperCase.contains(";FAN;DIR;")) {
+                updateFanDirection(messageParts);
+
+            } else if (messageUpperCase.contains(";FAN;AUTO;")) {
+                updateFanAuto(messageParts);
+
+            } else if (messageUpperCase.contains(";FAN;WHOOSH;STATUS;")) {
+                updateFanWhoosh(messageParts);
+
+            } else if (messageUpperCase.contains(";SMARTMODE;ACTUAL;")) {
+                updateFanSmartmode(messageParts);
+
+            } else if (messageUpperCase.contains(";FAN;SPD;MIN;")) {
+                updateFanSpeedMin(messageParts);
+
+            } else if (messageUpperCase.contains(";FAN;SPD;MAX;")) {
+                updateFanSpeedMax(messageParts);
+
+            } else if (messageUpperCase.contains(";LIGHT;PWR;")) {
+                updateLightPower(messageParts);
+
+            } else if (messageUpperCase.contains(";LIGHT;LEVEL;ACTUAL;")) {
+                updateLightLevel(messageParts);
+
+            } else if (messageUpperCase.contains(";LIGHT;AUTO;")) {
+                updateLightAuto(messageParts);
+
+            } else if (messageUpperCase.contains(";TBDSMARTERLIGHTING;")) {
+                // FIXME Still need to reverse engineer fan protocol to get this command format
```

Comment:
Unlikely.  Reverse engineering the protocol is a time-consuming process, considering I actually don't have one of these fans (someone who has one is helping me test).  I'll probably remove this until I can get a proper setup to discover the command format (i.e. my own fan and a sniffer for the wifi).